### PR TITLE
Tweak DPE patch path normalization behavior

### PR DESCRIPTION
**Pull request by NicholasVanSickle**
## What does this PR do?

I received some API feedback about patch handling within the Document Property Editor, namely that the prevalent use of `EndOfArray` entries in our `GenerateHierarchicalDeltaPatch` algorithm (and the validity of its prevalence) led to significant friction for consumers handling patches downstream, as to fully resolve an `EndOfArray` path requires a denormalization operation via a lookup in a provided `Dom::Value`, which is both cumbersome and potentially expensive. While there still remains a valid and open architectural question as to whether JSON patch is the right abstraction for the DPE, this patch attempts to address some of this friction in the short term.

To that end:
1. Our `EndOfArray` usage has been audited and restricted to only the one place where it's called out (and indeed mandated) in the JSON patch spec: `Add` operations, and `Add` operations only, denoting an insertion to the end of the target array.
2. An additional flag for patch generation has been added, `m_generateDenormalizedPaths`, in which we provide nonconformant paths for `Add` operations specifying an index past the end of the target array or node. This is a breaking change for serialized patches, but a non-breaking change for our internal patching algorithms, which have been softened to allow for this case.
3. Additionally, in the interest of allowing DPE consumers to continue to provide JSON patches in whatever form is most convenient, patches provided to an adapter are explicitly denormalized before being sent to any consumers (such as proxy adapters or the DPE widget itself).

The intent here is to make real, concrete, addressable paths *always* available to downstream systems to prevent unneeded and cumbersome book-keeping.

## How was this PR tested?

`Patch` unit tests still pass and I added a case for denormalizing an existing patch to prove out some of the scary bits about accepting arbitrary JSON patches. I also did some ad-hoc testing with the standalone DPE to validate basic patching behavior with no obvious regressions.

### CODE REVIEW

**Review by alexmontAmazon — APPROVED** [review]
Looks good! Very helpful, thanks!

File: Code/Framework/AzCore/AzCore/DOM/DomComparison.cpp [hunk 1]
```diff
@@ -101,8 +101,30 @@ namespace AZ::Dom
             {
                 if (i >= beforeSize)
                 {
-                    subPath.Push(PathEntry(PathEntry::EndOfArrayIndex));
-                    AddPatch(PatchOperation::AddOperation(subPath, after[i]), PatchOperation::RemoveOperation(subPath));
+                    // Per JSON patch RFFC 6902:
```
Comment (cgalvan) [inline Code/Framework/AzCore/AzCore/DOM/DomComparison.cpp:104]: Uber nit: 
```suggestion
                    // Per JSON patch RFC 6902:
```

**Review by cgalvan — APPROVED** [review]
Awesome thanks for doing this!

File: Code/Framework/AzCore/AzCore/DOM/DomPatch.h [hunk 2]
```diff
@@ -175,12 +184,40 @@ namespace AZ::Dom
...
+        //! This operation mutates the underyling patch operations; make a copy if you wish to keep the original patch.
+        //! \param rootElement The DOM element to patch.
+        //! \param strategy A callback to be run after every patch application, see PatchApplicationState.
+        //! \return an outcome with either the patched element or an error string
+        AZ::Outcome<Value, AZStd::string> ApplyAndDenormalize(
+            Value rootElement, StrategyFunctor strategy = PatchApplicationStrategy::HaltOnFailure);
+        //! Applies this patch to the given DOM element in place, changing it to the new value.
+        //! After applying the patch, any "EndOfArray" patch entries are denormalized into their resolved paths.
+        //! This operation mutates the underyling patch operations; make a copy if you wish to keep the original patch.
+        //! //! \param rootElement The DOM element to patch.
```
Comment (amzn-tmryan) [inline Code/Framework/AzCore/AzCore/DOM/DomPatch.h:209]: Super nitpick: extra //!

File: Code/Framework/AzCore/AzCore/DOM/DomPatch.h [hunk 3]
```diff
@@ -175,12 +184,40 @@ namespace AZ::Dom
         OperationsContainer::const_iterator cend() const;
         size_t size() const;
 
+        //! Applies this patch to the given DOM element.
+        //! \param rootElement The DOM element to patch.
+        //! \param strategy A callback to be run after every patch application, see PatchApplicationState.
+        //! \return an outcome with either the patched element or an error string
         AZ::Outcome<Value, AZStd::string> Apply(Value rootElement, StrategyFunctor strategy = PatchApplicationStrategy::HaltOnFailure) const;
+        //! Applies this patch to the given DOM element in place, changing it to the new value.
+        //! //! \param rootElement The DOM element to patch.
```
Comment (amzn-tmryan) [inline Code/Framework/AzCore/AzCore/DOM/DomPatch.h:193]: Super nitpick: extra //!

**Review by amzn-tmryan — APPROVED** [review]

**Review by dtamkin1 — APPROVED** [review]

File: Code/Framework/AzCore/AzCore/DOM/DomComparison.cpp (hunk 1 above)
Comment (NicholasVanSickle) [reply Code/Framework/AzCore/AzCore/DOM/DomComparison.cpp]: Fixed, thanks!

**Pull request merged**